### PR TITLE
feat(react-storage): skip adding too big files to upload queue

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

When file size is greater than 160 GB, prevent the file to be added to the upload view and render a warning message. Users can ignore the message and proceed the upload, the message will be automatically dismissed when the upload starts.
Screenshot 2024-11-17 at 6 38 24 PM

(Reduced the size limit to render the warning message)
Screenshot 2024-11-15 at 4 45 25 PM

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [ ] `yarn test` passes and tests are updated/added
- [ ] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts [hunk 1]
```diff
@@ -79,6 +80,19 @@ export const DEFAULT_UPLOAD_VIEW_DISPLAY_TEXT: DefaultUploadViewDisplayText = {
...
+  getFilesValidationMessage: ({ invalidFiles } = {}) => {
+    if (!invalidFiles) {
+      return undefined;
+    }
+    const fileNames = invalidFiles
+      .filter(({ file }) => file.size > UPLOAD_FILE_SIZE_LIMIT)
+      .map(({ file }) => file.name)
+      .join(', ');
+    return {
+      content: `File ${fileNames} exceeds the size limit 160 GB. Failed to add to upload queue.`,
```
Comment (HuiSF) [inline packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts:92]: If I understood correctly the 160GB limit is per file? Can the message be something like "These files cannot be added to the upload queue due to they are larger than 160GB respectively" could be clearer?

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts [hunk 2]
```diff
@@ -3,30 +3,58 @@ import React from 'react';
...
 
+  const validFiles = React.useMemo(
+    () =>
+      files?.filter((fileItem) => {
+        const { id, file } = fileItem;
+        if (file.size > UPLOAD_FILE_SIZE_LIMIT) {
+          setInvalidFiles((prev) =>
+            isUndefined(prev) ? [fileItem] : prev.concat(fileItem)
+          );
+          dispatchStoreAction({ type: 'REMOVE_FILE_ITEM', id });
```
Comment (HuiSF) [inline packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts:36]: Nit: I'd avoid attach a side effect into the filter callback function.

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx [hunk 3]
```diff
@@ -73,6 +75,8 @@ export function UploadView({
     ? getActionCompleteMessage({
         counts: statusCounts,
       })
+    : invalidFiles && !isProcessing
+    ? getFilesValidationMessage({ invalidFiles })
```
Comment (HuiSF) [inline packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx:79]: Nit: now we are expanding to nested ternary expressions it could be extracted into a util function with clearer if flows. :D

File: packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts [hunk 4]
```diff
@@ -79,6 +80,19 @@ export const DEFAULT_UPLOAD_VIEW_DISPLAY_TEXT: DefaultUploadViewDisplayText = {
 
     return { content: 'All files uploaded.', type };
   },
+  getFilesValidationMessage: ({ invalidFiles } = {}) => {
+    if (!invalidFiles) {
+      return undefined;
+    }
+    const fileNames = invalidFiles
```
Comment (HuiSF) [inline packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts:87]: Sanity check: invalidFiles can be an empty array here which could generate an incomplete error message, will this be expected?

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts:36]: Taking a closer look at this one and I realize if we have multiple files larger than 160 GB, the side effects `setInvalidFiles` and `dispatchStoreAction` will not work as we wanted. So instead of supporting adding files to the file context and remove invalid ones in this hook, we should prevent adding these invalid files to the state in the 1st place.

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx (hunk 3 above)
Comment (AllanZhengYP) [reply packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx:79]: Good catch... it's a long day for me😵‍💫

File: packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts (hunk 4 above)
Comment (AllanZhengYP) [reply packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts:87]: It cannot be empty array according to how we populate the array, but I will handle the scenario here as well.

**Review by HuiSF — APPROVED** [review]

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx [hunk 5]
```diff
@@ -68,12 +70,15 @@ export function UploadView({
...
-
-  const message = isProcessingComplete
+  const actionCompleteMessage = isProcessingComplete
     ? getActionCompleteMessage({
         counts: statusCounts,
       })
     : undefined;
+  const filesValidationMessage =
+    invalidFiles && !isProcessing
+      ? getFilesValidationMessage({ invalidFiles })
```
Comment (pranavosu) [inline packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx:80]: i'm not certain about this but is it better to just have a single `getMessage` function that takes in all the variables and returns a single message as a pattern. I think I remember @calebpollman mentioning this somewhere. 

something like 

```
const message = getMessage({
      isProcessingComplete,
      invalidFiles,
      isProcessing,
      counts
})
```

**Review by HuiSF — APPROVED** [review]

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx (hunk 5 above)
Comment (AllanZhengYP) [reply packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx:80]: A single `getMessage` would have to accept a lot view states to the callback. It would be more susceptible to breaking changes. I think it's cleaner to expose less view states to the displayText. 

Later we should refactor the message component itself to support multiple messages instead.

Comment (pranavosu) [reply packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/UploadView.tsx:77]: Yeah, coupling these would make changes more difficult but the display logic is hard to follow right now, so that's the tradeoff. i think we can handle it in a follow up

**Review by pranavosu — APPROVED** [review]

File: packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts [hunk 6]
```diff
@@ -79,6 +80,19 @@ export const DEFAULT_UPLOAD_VIEW_DISPLAY_TEXT: DefaultUploadViewDisplayText = {
...
     return { content: 'All files uploaded.', type };
   },
+  getFilesValidationMessage: ({ invalidFiles = [] } = {}) => {
+    const tooBigFileNames = invalidFiles
+      .filter(({ file }) => isFileTooBig(file))
+      .map(({ file }) => file.name)
+      .join(', ');
+    if (tooBigFileNames) {
+      return {
+        content: `These files cannot be added to the upload queue due to they are larger than 160GB respectively: ${tooBigFileNames}`,
```
Comment (pranavosu) [inline packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts:90]: nit
```suggestion
        content: `These files cannot be added to the upload queue as they are larger than 160GB: ${tooBigFileNames}`,
```

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts [hunk 7]
```diff
@@ -3,30 +3,57 @@ import React from 'react';
...
+          isUndefined(prev) ? [fileItem] : prev.concat(fileItem)
+        );
+        dispatchStoreAction({ type: 'REMOVE_FILE_ITEM', id });
+      } else {
+        validFileItems.push(fileItem);
+      }
+    });
+
+    return validFileItems;
+  }, [files, dispatchStoreAction]);
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts:41]: We can remove the side effect behaviors by returning both valid files and invalid files here:
```
const { validFiles, invalidFiles } = React.useMemo(() => ..);
```

We can clean out the invalid files from the store by exposing an `onMessageDismiss` callback on the return of `useUploadView`, something like:
```
const onMessageDismiss = () => {
  invalidFiles?.map((: data }) =>  dispatchStoreAction({ type: 'REMOVE_FILE_ITEM', id: data.id });
}
```

And provide it to the controls context when `invalidFiles` is truthy

File: packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts [hunk 8]
```diff
@@ -79,6 +80,19 @@ export const DEFAULT_UPLOAD_VIEW_DISPLAY_TEXT: DefaultUploadViewDisplayText = {
...
     return { content: 'All files uploaded.', type };
   },
+  getFilesValidationMessage: ({ invalidFiles = [] } = {}) => {
+    const tooBigFileNames = invalidFiles
+      .filter(({ file }) => isFileTooBig(file))
+      .map(({ file }) => file.name)
+      .join(', ');
+    if (tooBigFileNames) {
+      return {
+        content: `Files larger that 160GB cannot be added to the upload queue: ${tooBigFileNames}`,
```
Comment (pranavosu) [inline packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts:90]: my comment had a typo :(

```suggestion
        content: `Files larger than 160GB cannot be added to the upload queue: ${tooBigFileNames}`,
```

**Review by pranavosu — CHANGES_REQUESTED** [review]

File: packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts (hunk 7 above)
Comment (AllanZhengYP) [reply packages/react-storage/src/components/StorageBrowser/views/LocationActionView/UploadView/useUploadView.ts:41]: However this will make the invalid files exist in the files context longer than we want. Currently they are removed when the upload started. Now we want it to be removed only when the message is dismissed?

File: packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts [hunk 9]
```diff
@@ -79,6 +80,19 @@ export const DEFAULT_UPLOAD_VIEW_DISPLAY_TEXT: DefaultUploadViewDisplayText = {
 
     return { content: 'All files uploaded.', type };
   },
+  getFilesValidationMessage: ({ invalidFiles = [] } = {}) => {
+    const tooBigFileNames = invalidFiles
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/displayText/libraries/en/uploadView.ts:84]: Prefer early return when `invalidFiles` is undefined

**Review by calebpollman — APPROVED** [review]
Minor feedback, non-blocking
